rfcs/3-add-event-notification-flags.md

```diff
@@ -0,0 +1,232 @@
+infrastructure doesn't allow to simply allow for a new flag to be
+specified by the user: the JavaScript API has to be changed, there might
+be implications to the backwards compatibility depending on the
+implementation, and the implementation itself may vary based on the
+requirements: for example, we might also stop ignoring the event
+command name and also expose it for the user to subscribe.
+
+# Suggested Solutions
+
+
+## [Option 1](https://github.com/RedisGears/RedisGears/pull/1061)
+
+Use the existing API of `redis.registerKeySpaceTrigger` and extend the
+optional dictionary argument to also accept a list of event notification
+flags, which the user wants to subscribe to. For this to happen, the
+event notification flags must also be known by the user and easily
+accessible in the JavaScript code, so an enum containing the event
+notification flag values is exposed. The minor version of the JavaScript
+API is bumped, what means full backwards compatibility: if the flags are
+not specified by the user in the optional dictionary argument, the
+behaviour of the older (current) JavaScript API remains, leaving to
+exactly the same results as prior to the change.
+
+
+Examples:
+
+
+```JS
+#!js name=lib api_version=1.2
+
+redis.registerKeySpaceTrigger(
+  'onlynewnotifications', // trigger name
+  '', //key prefix
+  function(client, data) {
+    console.log("Got this key data created1: " + data);
+  }, //callback
+  {
+    description: 'description',
+    onTriggerFired: function(client, data) {
+        console.log("Got this key data created2: " + data);
+    },
+    // The new optional argument to specify the event notification flags:
+    eventNotificationFlags: [EventNotificationFlags.NEW]
+  } //optional arguments
+)
+
+redis.registerKeySpaceTrigger(
+  'allnewnotifications', // trigger name
+  '', //key prefix
+  function(client, data) {
+    console.log("Got this key data 1: " + data);
+  }, //callback
+  {
+    description: 'description',
+    onTriggerFired: function(client, data) {
+        console.log("Got this key data 2: " + data);
+    },
+    // The new optional argument to specify the event notification flags:
+    eventNotificationFlags: [EventNotificationFlags.NEW, EventNotificationFlags.ALL]
+  } //optional
+)
+```
+
+This has an advantage of solving the problem, but if we already go this
+way to expose the flags, we might need to filter some of those values
+in the original Redis enumeration variants, which we shouldn't expose
+to the user.
+
+Another drawback is that JavaScript doesn't have any means to check the
+values for incompatibility with each other or any other sort of overlaps.
+This will have to be done at deploy time, by RedisGears itself.
+
+## Option 2 (Expose the event command)
+
+However, if we already think about the first approach, we might as well
+reconsider the event command argument of the notification, and expose it
+too; just as well as we could allow the user's key space trigger to
```

Comment:
Do we know about real world use cases that will benefit from exposing the command name?
How do we expect to handle renamed commands here?

Reply:
Unfortunately, I know nothing about the real-world use cases. @MeirShpilraien @thomascaudron Do you know, perhaps?
Handling renamed commands should be done by the user in this case, not by our side: we simply pass through the event command name string to the user.

Reply:
I think there is a confusion. Its not really the command name, its the event name that was fire by Redis (which IIRC is not effected by command renaming).
I do not know about usecases but I believe its an important information to expose (we already expose it today).